### DISCUSSION TITLE
temporary mod to automatically adjust offset after 100 hit objects

### DISCUSSION THREAD

**Discussion by Doksha** [ppy/osu#20862] [Ideas]
lazer has a huge issue with offset, every single map needs a different offset (usually +- 10ms, or a total range of 20ms, not considering mistimed maps, and around +- 20ms if you consider mods, which also seem to affect offset), which can make playing multi's pretty annoying since you cant restart to adjust the offset.
Until the actual underlying issue is fixed, maybe have a temporary mod that adjusts the offset as necessary after the first 100 hit objects. This mod wouldnt be implemented in the full game and would just be a bandaid fix.

This mod would only apply to local/multi scores if possible, global scores would be unaffected

**Comment by WindowsMeosu** [comment]
I feel like <b>'every 100 hit objects and every BPM change</b>' would work better. "every BPM change" since there are mistimed maps with multiple BPM changes out there.

**Reply by Doksha** [reply]
I've been thinking about it, and probably the best option is to set the offset to the inverse of the average of the early/late timings from the past 100 objects (with perfectly timed hits assumed for before the map starts)

But that would only be if this were to be a full game feature, I more just suggested this as a temporary fix for lazer's awful offset issue.

**Comment by frenzibyte** [comment]
I don't understand the point of this at all. Isn't it just a duplicate of the existing beatmap offset wizard in player loading session? I don't think it would be nice for the player to suddenly play at a different offset midway through the beatmap.

> lazer has a huge issue with offset, every single map needs a different offset...

For clarification, is this a lazer-only issue? If so, it would be much better opening an issue thread about it with proper and correct details for further investigation, a discussion thread, or even bumping an existing issue thread if it's present on ranked beatmaps. All rather than proposing bandaids here.

> and around +- 20ms if you consider mods, which also seem to affect offset

Are you sure about this? any mod which affect rate also applies certain adjustment to make it not affect the offset as well, as long as you applied a correct offset on the beatmap itself.

**Reply by Doksha** [reply]
the existing offset adjustment only applies after you restart, it doesnt apply mid-map, meaning you can't adjust offset in multiplayer or marathon maps. Hardrock and halftime definitely have a huge effect (halftime has always had this issue in stable, but hardrock is fixed in stable).
I dont want to have to restart every map midway through just to get the offset right.
I have only tested this issue with ranked maps (it happens even on reputable/farm mappers, so its definitely not mapper error) but I have not compared each map to stable.
I cant seem to find an existing issue, so if you can confirm there isnt one I will create it.

My updated proposal (the reply to [WindowsMeosu](https://github.com/WindowsMeosu)) also makes this mod useful for basically every map that has changing offset throughout the map. Its just QOL, even if it wont be necessary for ranked maps, its still very useful for mistimed maps.

**Reply by quat1024** [reply]
Did you mean doubletime? I would be very surprised if hardrock actually has an effect on offset other than "reacting to ar10 is hard"

**Reply by integerrr** [reply]
This proposal makes even less sense then if the offset isn't applied mid-map. If your problem is finding it annoying having to restart a map to calibrate your offset because the data is lost when exiting gameplay/finishing a map, I'd much rather see the calibration taps from the offset wizard in player loading screen persist outside of gameplay than whatever this is made into a mod.

**Reply by quat1024** [reply]
It sounds to me sortof like Stepmania's AutoSync feature, which

> I don't think it would be nice for the player to suddenly play at a different offset midway through the beatmap.

does exactly that, every couple dozen arrows. It's not intended as a gameplay modifier of course, but when getting an initial sync for a chart or for your computer I think the immediate feedback is pretty handy, it's a lot quicker than play -> get a feel for the offset -> restart -> wait -> play -> get a feel for the offset -> restart -> wait etc

**Reply by Doksha** [reply]
> Did you mean doubletime? I would be very surprised if hardrock actually has an effect on offset other than "reacting to ar10 is hard"

I consistently hit on time with hardrock in stable after warming up, but in lazer its always off (I cant remember how much before/after) I also get decent UR so its not due being unable to react to AR 10, although it might be due to AR 10 being tweaked in stable and not in lazer (I do not know if this is this is the case, just speculation)

**Reply by Doksha** [reply]
> does exactly that, every couple dozen arrows. It's not intended as a gameplay modifier of course, but when getting an initial sync for a chart or for your computer I think the immediate feedback is pretty handy, it's a lot quicker than play -> get a feel for the offset -> restart -> wait -> play -> get a feel for the offset -> restart -> wait etc

I was basically thinking that, although a little more gradual. keep a buffer of the last 100 hits early / late timing, and adjust the offset to the average of that. Would make the change relatively unnoticable while fixing offset
